tests/test_responsetypes.py

```diff
@@ -72,17 +72,39 @@ def test_from_headers(self):
             retcls = responsetypes.from_headers(source)
             assert retcls is cls, f"{source} ==> {retcls} != {cls}"
 
-    def test_from_args(self):
-        # TODO: add more tests that check precedence between the different arguments
+    def test_from_args_pre_xtractmime(self):
         mappings = [
             ({'url': 'http://www.example.com/data.csv'}, TextResponse),
-            # headers takes precedence over url
             ({'headers': Headers({'Content-Type': ['text/html; charset=utf-8']}),
               'url': 'http://www.example.com/item/'}, HtmlResponse),
             ({'headers': Headers({'Content-Disposition': ['attachment; filename="data.xml.gz"']}),
               'url': 'http://www.example.com/page/'}, Response),
+            ({'body': b'Some plain\0 text data with\0 tabs and null bytes\0'}, TextResponse),
+            ({'body': b'\x03\x02\xdf\xdd\x23', 'headers': Headers({'Content-Encoding': 'UTF-8'})},
+             Response),
+            ({'body': b'\x00\x01\xff', 'url': '://www.example.com/item/',
+              'headers': Headers({'Content-Type': b'text/plain'})}, TextResponse),
+            ({'url': 'http://www.example.com/item/file.html'}, HtmlResponse),
+            ({'body': b'<html><head><title>Hello</title></head>'}, HtmlResponse),
+            ({'body': b'<?xml version="1.0" encoding="utf-8"'}, XmlResponse),
+            ({'filename': 'file.pdf'}, Response),
+            ({'url': 'http://www.example.com/item/file.pdf'}, Response),
+            ({'body': b'Some plain text data\1\2 with tabs and\n null bytes\0'}, Response),
+            ({'filename': '/tmp/temp^'}, TextResponse),
+        ]
+        for source, cls in mappings:
+            retcls = responsetypes.from_args(**source)
+            assert retcls is cls, f"{source} ==> {retcls} != {cls}"
 
-
+    def test_from_args_post_xtractmime(self):
+        mappings = [
+            # different behaviour with http and non-http urls
+            ({'body': b'\x00\x01\xff', 'url': 'http://www.example.com/item/',
+              'headers': Headers({'Content-Type': b'text/plain'})}, Response),
+            ({'body': b'%PDF-1.4'}, Response),
+            ({'body': b'%PDF-1.4', 'headers': Headers({'Content-Type': b'application/pdf'})}, Response),
+            ({'headers': Headers({'Content-Type': b'application/ecmascript'})}, TextResponse),
+            ({'headers': Headers({'Content-Type': b'application/ld+json'})}, TextResponse),
```

Comment:
I’ve reviewed the previous implementation and taken some notes for you to review, and check if there are behavioral changes and, if so, add test cases for them:

- Before, a `Content-Encoding` header would always cause a `Response`. I imagine now that’s not always the case.
  In fact, I suspect a non-binary body reporting some encoding before would cause the HTTPCompressionMiddleware to raise an exception, but with the new implementation the middleware should do nothing if the response is not recognized as binary.
- Does the new implementation work with these types from `CLASSES`?
  ```
  'application/x-json': 'scrapy.http.TextResponse',
  'application/json-amazonui-streaming': 'scrapy.http.TextResponse',
  'application/x-javascript': 'scrapy.http.TextResponse',
  ```
- Before, a `Content-Type` missing or for a binary type would be ignored in favor of `Content-Disposition`, URL or file name. I am guessing, for example, that with `Content-Type` `application/octet-stream` but a path-like field of a non-binary type, before you would get a non-`Response` class, and now you get `Response`.
- When inspecting the body:
    - `0x0C` and `0x1B` were considered binary, and now are not (the standard does not include them in the binary data range set).
    - `<html>` anywhere in the first 5000 bytes would trigger HtmlResponse, even if the body was: `this is not <html>`
    - `<?xml` anywhere in the first 5000 bytes would trigger XmlResponse, even if the body was: `this is not <?xml either`

Reply:
> Before, a Content-Encoding header would always cause a Response. I imagine now that’s not always the case.
In fact, I suspect a non-binary body reporting some encoding before would cause the HTTPCompressionMiddleware to raise an exception, but with the new implementation the middleware should do nothing if the response is not recognized as binary.

Maybe I am testing wrong, but both of these testcases are passing with older implementation

 ```({'body': b'Some binary\1\2 text', 'headers': Headers({'Content-Encoding': 'UTF-8'})},
             Response),

 ({'body': b'Some plain text', 'headers': Headers({'Content-Encoding': 'UTF-8'})},
             TextResponse),```

Reply:
>     * Does the new implementation work with these types from `CLASSES`?
>       ```
>       'application/x-json': 'scrapy.http.TextResponse',
>       'application/json-amazonui-streaming': 'scrapy.http.TextResponse',
>       'application/x-javascript': 'scrapy.http.TextResponse',
>       ```

Nope. I think we can hard-code them while detecting a text response in `_guess_response_type`??

Reply:
> * Before, a `Content-Type` missing or for a binary type would be ignored in favor of `Content-Disposition`, URL or file name. I am guessing, for example, that with `Content-Type` `application/octet-stream` but a path-like field of a non-binary type, before you would get a non-`Response` class, and now you get `Response`.
> 
> * When inspecting the body:
>   
>   * `0x0C` and `0x1B` were considered binary, and now are not (the standard does not include them in the binary data range set).
>   * `<html>` anywhere in the first 5000 bytes would trigger HtmlResponse, even if the body was: `this is not <html>`
>   * `<?xml` anywhere in the first 5000 bytes would trigger XmlResponse, even if the body was: `this is not <?xml either`

👍🏼